academy/xl-cosmjs/with-kepler.md

```diff
@@ -0,0 +1,496 @@
+        <fieldset className={styles.card}>
+            <legend>Send</legend>
+            <p>To faucet:</p>
+            <input value={toSend} type="number" onChange={this.onToSendChanged}/> {denom}
+            <button onClick={this.onSendClicked}>Send to faucet</button>
+        </fieldset>
+      </div>
+    }
+}
+```
+
+</ExpansionPanel>
+
+Of note is that:
+
+* Its **properties** happen to only contain the things it knows at build time.
+* It keeps a **state**, and this state is either updated by the user or will be updated after a fetch.
+* It reuses some default style you can find in `/styles`.
+
+The component is still unused. Put it inside `index.tsx` like so:
+
+```typescript
+const Home: NextPage = () => {
+  return <FaucetSender
+    faucetAddress="cosmos1umpxwaezmad426nt7dx3xzv5u0u7wjc0kj7ple"
+    rpcUrl="http://198.50.215.1:46657" />
+}
+```
+
+When `npm run dev` picks up the changes, you should see that your page has changed to what you created.
+
+It is not very useful just yet, but you can make it more so.
+
+## Install CosmJs
+
+Now that you have a working Next.js project and ready page, it is time to add the necessary CosmJs elements to the project:
+
+```sh
+$ npm install @cosmjs/stargate cosmjs-types --save
+```
+
+## Show what can be shown
+
+When building a GUI, it is good practice to not ask your user's address until the point where you have presumably gained their trust and / or it is necessary, e.g. if they clicked a relevant button. So you ought to start by showing information that is knowable without user input. Here, it is `denom` and the faucet balance. Add a function for that:
+
+```typescript
+updateFaucetBalance = async(client: StargateClient) => {
+    const balances: readonly Coin[] = await client.getAllBalances(this.props.faucetAddress)
+    const first: Coin = balances[0]
+    this.setState({
+        denom: first.denom,
+        faucetBalance: first.amount,
+    })
+}
+```
+
+Notice how it only cares about the first coin type. Add that in the constructor as well so that it does it on load. Via another specific function:
+
+```typescript
+constructor(props:FaucetSenderProps) {
+    ...
+    setTimeout(this.init, 500)
+}
+
+init = async() => this.updateFaucetBalance(await StargateClient.connect(this.props.rpcUrl))
+```
+
+After `run dev` picks the changes, you should see your page that starts showing relevant information.
+
+Now, add elements that will handle your user's information.
+
+## Detect Keplr
+
+Following [Keplr's documentation](https://docs.keplr.app/api/#how-to-detect-keplr), it is time to add a function to see if Keplr is installed on the browser. For convenience and type hinting, install the Typescript Keplr types:
+
+```sh
+$ npm install @keplr-wallet/types --save-dev
+```
+
+And inform Typescript that `window` may have a `.keplr` field along with [this helper](https://github.com/chainapsis/keplr-wallet/tree/master/docs/api#keplr-specific-features), by adding it to `FaucetSender.tsx`:
```

Comment:
```suggestion
And inform Typescript that `window` may have a `.keplr` field with the help of [this helper](https://github.com/chainapsis/keplr-wallet/tree/master/docs/api#keplr-specific-features), by adding it to `FaucetSender.tsx`:
```
@rob-Hitchens I originally wanted to say that the helper informs Typescript, but it was poorly worded.